### PR TITLE
[AC-1904] Implement endpoint to retrieve Provider subscription

### PR DESCRIPTION
**Pull request by amorask-bitwarden**
## Type of change

```
- [ ] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

This PR is part of the Consolidated Billing project. It wires up a new endpoint at `/api/providers/{providerId}/billing/subscription` that allows a client to retrieve information about a provider's Stripe subscription and current plan configurations.

This PR is based off of https://github.com/bitwarden/server/pull/3875, which is currently in QA.

https://github.com/bitwarden/server/assets/144709477/3f1cccdf-b19b-4c46-8f5f-30d1a573bcda

### API
* **[/AdminConsole/Controllers/OrganizationsController.cs](https://github.com/bitwarden/server/pull/3913/files#diff-a1fca15df7919857ea1493710aad9fbd529e1a5c623db0f6573412e04f57becb):** Renamed injected dependency.
* **[/Auth/Controllers/AccountsController.cs](https://github.com/bitwarden/server/pull/3913/files#diff-19ca62a64f34d8a3f432d8d76e1f578f5dc87d1ff2859eb789faf6ae41fb1c20):** Renamed injected dependency.
* **[/Billing/Controllers/ProviderBillingController.cs](https://github.com/bitwarden/server/pull/3913/files#diff-fa8e4e1e293b4d5e549ed378829bd9b72694a3d8bb72be22547480a9f92e0823):** Created controller and exposed new endpoint that returns `ProviderSubscriptionDTO`.
* **[/Billing/Models/ProviderSubscriptionDTO.cs](https://github.com/bitwarden/server/pull/3913/files#diff-07d6696466aa348e3963c01d6f2c4ef2e68a4a563ae9ad837d0ce124cd097d8d):** Created DTO representing a provider subscription.

### Core
* **[/AdminConsole/Entities/Provider/Provider.cs](https://github.com/bitwarden/server/pull/3913/files#diff-cbc5e8e8d3e0a475872658b0729ee8d05e8fffa0b0b8da0c8af963163e8dca1a):** Updated entity to implement `ISubscriber`.
* **[/Billing/BillingException.cs](https://github.com/bitwarden/server/pull/3913/files#diff-a3bc90beeca04cfd3bb665d3a0542bdda54049af945fa52598a148165128f18b):** Created new exception type for Billing.
* **[/Billing/Commands/ICancelSubscriptionCommand.cs](https://github.com/bitwarden/server/pull/3913/files#diff-8b1458c81ac2ee646ea698e3224fd126c9d8d9aab23d3018c8f1ca52b644fb6f):** Removed exception annotation as these are difficult to manage consistently.
* **[/Billing/Commands/IRemovePaymentMethodCommand.cs](https://github.com/bitwarden/server/pull/3913/files#diff-9b466dddca300bf3a4977370a880bfc165fa47eaae6ed2df60aad0f0259dfc8e):** Added code comments.
* **[/Billing/Commands/Implementations/RemovePaymentMethodCommand.cs](https://github.com/bitwarden/server/pull/3913/files#diff-d8ffb290e5147ba2381b02917aa055ac6a0b951e216113ccf20737b29e4f2107):** Refactored to primary constructor.
* **[/Billing/Entities/ProviderPlan.cs](https://github.com/bitwarden/server/pull/3913/files#diff-b5ae99f4cb315ffda59d280b5764f8fb3e25fb8fe0d2db073bbb5a51bde756b2):** Removed unused column/property - will be deleted as part of https://bitwarden.atlassian.net/browse/AC-2311.
* **[/Billing/Extensions/ServiceCollectionExtensions.cs](https://github.com/bitwarden/server/pull/3913/files#diff-988e374c5d5e653d6e1cf10aa61ccf6124d7181d3fbae93d27bc1d4c1c68014b):** Added `ProviderBillingQueries` to registration extension method. 
* **[/Billing/Models/ConfiguredProviderPlan.cs](https://github.com/bitwarden/server/pull/3913/files#diff-3e5e3b0bbb4984f6ef90d2524c74282441111830cc53dea24181f3e5fcb635cb):** Created model representing a configured provider plan.
* **[/Billing/Models/ProviderSubscriptionData.cs](https://github.com/bitwarden/server/pull/3913/files#diff-02f8ce6f58528103f1110f6c569e6fd3e54aa96829dc63534eb74a4ff6d78383):** Created model representing provider subscription data.
* **[/Billing/Queries/Implementations/ProviderBillingQueries.cs](https://github.com/bitwarden/server/pull/3913/files#diff-7e0b33fceef8d16f9dd08e89754f3dc65323449f427a13d4efa6977742f7f068):** Created class containing provider billing queries.
* **[/Billing/Queries/Implementations/SubscriberQueries.cs](https://github.com/bitwarden/server/pull/3913/files#diff-1aecf3cc7b0dc01fe2080162f27614f1ab64880b8207a5e129eaca59c1d7fa83):** Renamed from `GetSubscriptionQuery`.
* **[/Billing/Utilities.cs](https://github.com/bitwarden/server/pull/3913/files#diff-53decb093838402e409962f65ca70b1cada998c47a3eaf04ad54599a632a6433):** Updates utility method to use `BillingException`. 
* **[/Constants.cs](https://github.com/bitwarden/server/pull/3913/files#diff-94998af117f486ce85a16234b1bf36dd8c918651ff4d5c851927d203f5120e4e):** Added feature flag.

### Infrastructure.Dapper
* **[/Billing/Repositories/ProviderPlanRepository.cs](https://github.com/bitwarden/server/pull/3913/files#diff-32a1fc202299e351238309606c796828dee1218b2b7f1cf4e3c715d8aa0d8531):** Fixed return type of `GetByProviderId`.

### Infrastructure.EntityFramework
* **[/Billing/Repositories/ProviderPlanRepository.cs](https://github.com/bitwarden/server/pull/3913/files#diff-459423670a37318ecfadbf7883c551de6814c9f5dde85fdee605dfcb810c3f27):** Fixed return type of `GetByProviderId`.

## Before you submit

- Please check for formatting errors (`dotnet format --verify-no-changes`) (required)
- If making database changes - make sure you also update Entity Framework queries and/or migrations
- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

File: src/Core/Billing/Queries/Implementations/ProviderBillingQueries.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        if (subscription == null)
+        {
+            return null;
+        }
+
+        var providerPlans = await providerPlanRepository.GetByProviderId(providerId);
+
+        var configuredProviderPlans = providerPlans
+            .Where(providerPlan => providerPlan.Configured)
+            .Select(ConfiguredProviderPlan.From)
```
Comment (cturnbull-bitwarden) [inline src/Core/Billing/Queries/Implementations/ProviderBillingQueries.cs:42]: Nice, I like this pattern for mapping.

File: src/Core/Billing/Utilities.cs [hunk 2]
```diff
@@ -1,8 +1,11 @@
...
-
-namespace Bit.Core.Billing;
+﻿namespace Bit.Core.Billing;
 
 public static class Utilities
 {
-    public static GatewayException ContactSupport() => new("Something went wrong with your request. Please contact support.");
+    public const string BraintreeCustomerIdKey = "btCustomerId";
+
+    public static BillingException ContactSupport(
```
Comment (cturnbull-bitwarden) [inline src/Core/Billing/Utilities.cs:7]: 💯 Excellent

**Review by cturnbull-bitwarden — COMMENTED** [review]
This is great work. Thank you, sir!

**Review by JaredSnider-Bitwarden — APPROVED** [review]
Auth changes look good. Did not review rest of PR.

**Review by addisonbeck — APPROVED** [review]
AC changes look good.

**Review by cturnbull-bitwarden — APPROVED** [review]

**Pull request merged**